File: polyglot/java/emulator_avd_spoof_detector.java
package polyglot.java;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Emulator AVD Spoof Detector - Detects Android Virtual Device (AVD) environments.
 * 
 * Checks multiple vectors: build props, env vars, file artifacts, processes, network interfaces.
 * Returns a confidence score and detailed findings.
 */
public final class EmulatorAVDSpoofDetector {

    private static final int EMULATOR_SCORE_THRESHOLD = 3;
    
    /**
     * Main detection entry point.
     * @return DetectionResult with score, verdict, and findings
     */
    public static DetectionResult detect() {
        return detect(System.getenv());
    }

    /**
     * Detect using provided environment map (useful for unit tests).
     */
    public static DetectionResult detect(Map<String, String> env) {
        List<Finding> findings = new ArrayList<>();
        
        // 1. Build Properties Analysis
        findings.addAll(checkBuildProperties(env));
        
        // 2. Environment Variable Analysis
        findings.addAll(checkEnvironmentVariables(env));
        
        // 3. File System Artifacts (if running on Android)
        findings.addAll(checkFileArtifacts());
        
        // 4. Process Detection
        findings.addAll(checkProcesses());
        
        // 5. Network Interface Analysis
        findings.addAll(checkNetworkInterfaces());
        
        // 6. Sensor/Device Pattern Analysis
        findings.addAll(checkSensorPatterns(env));
        
        int score = calculateScore(findings);
        return new DetectionResult(score, findings);
    }

    /**
     * Calculate confidence score based on findings.
     */
    private static int calculateScore(List<Finding> findings) {
        int score = 0;
        for (Finding f : findings) {
            if (!f.isSoft()) {
                score += f.getWeight();
            }
        }
        return Math.min(score, 10); // Cap at 10
    }

    /**
     * Determine verdict from score.
     */
    private static String getVerdict(int score) {
        if (score >= EMULATOR_SCORE_THRESHOLD) {
            return "LIKELY_EMULATOR";
        } else if (score > 0) {
            return "POSSIBLE_EMULATOR";
        }
        return "REAL_DEVICE";
    }

    /**
     * Check Android build properties for emulator signatures.
     */
    private static List<Finding> checkBuildProperties(Map<String, String> env) {
        List<Finding> findings = new ArrayList<>();
        
        // Common emulator product model patterns
        Set<String> emulatorModels = Set.of(
            "Android SDK Built-in",
            "Android SDK Phone",
            "emulator-5554",
            "sdk_phone_x86",
            "Generic Android TV"
        );
        
        String model = env.get("ro.product.model");
        if (model != null && emulatorModels.contains(model)) {
            findings.add(new Finding(
                "EMULATOR_MODEL_SIGNATURE",
                "Product model matches known AVD template: " + model,
                5, false));
        }
        
        // Check for SDK-specific build props
        String sdkVersion = env.get("ro.build.version.sdk");
        if (sdkVersion != null) {
            int sdkInt;
            try {
                sdkInt = Integer.parseInt(sdkVersion);
                // AVD often has specific SDK patterns or "release" strings
                if (sdkInt < 21 || sdkInt > 35) {
                    findings.add(new Finding(
                        "UNUSUAL_SDK_RANGE",
                        "SDK version outside typical range: " + sdkVersion,
                        2, true));
                }
            } catch (NumberFormatException e) {
                // Non-numeric SDK - might be emulator-specific string
                if (!sdkVersion.contains("release")) {
                    findings.add(new Finding(
                        "UNUSUAL_SDK_FORMAT",
                        "Non-standard SDK format: " + sdkVersion,
                        2, true));
                }
            }
        }
        
        // Check for emulator-specific build fingerprint patterns
        String fingerprint = env.get("ro.build.fingerprint");
        if (fingerprint != null) {
            if (fingerprint.contains("_emulator") || 
                fingerprint.contains("/emulator/") ||
                fingerprint.contains("sdk_phone")) {
                findings.add(new Finding(
                    "EMULATOR_FINGERPRINT",
                    "Fingerprint contains emulator marker: " + fingerprint,
                    4, false));
            }
        }
        
        // Check for Android SDK specific props
        String sdkPath = env.get("ro.sdk.path");
        if (sdkPath != null && sdkPath.contains("AndroidSdk")) {
            findings.add(new Finding(
                "SDK_PATH_ARTIFACT",
                "SDK path suggests AVD: " + sdkPath,
                2, true));
        }
        
        return findings;
    }

    /**
     * Check environment variables for emulator indicators.
     */
    private static List<Finding> checkEnvironmentVariables(Map<String, String> env) {
        List<Finding> findings = new ArrayList<>();
        
        // Android-specific emulator env vars
        Set<String> emulatorEnvVars = Set.of(
            "ANDROID_SERIAL",
            "ANDROID_VNDK_ROOT",
            "ANDROID_HOME",
            "ANDROID_SDK_ROOT",
            "ANDROID_NDK_ROOT"
        );
        
        for (String key : emulatorEnvVars) {
            String value = env.get(key);
            if (value != null && !value.isEmpty()) {
                findings.add(new Finding(
                    "EMULATOR_ENV_VAR_" + key,
                    "Environment variable present: " + key + "=" + value,
                    2, true));
            }
        }
        
        // Check for QEMU-specific env vars (common in AVD)
        if (env.containsKey("QEMU_AUDIO_DRV") || 
            env.containsKey("SDL_AUDIODRIVER")) {
            findings.add(new Finding(
                "QEMU_AUDIO_DRIVER",
                "QEMU/SDL audio driver detected: " + env.get("QEMU_AUDIO_DRV"),
                3, false));
        }
        
        // Check for virtual device specific vars
        if (env.containsKey("ANDROID_VNDK_ROOT")) {
            findings.add(new Finding(
                "VNDK_ARTIFACT",
                "Android VNDK root set: " + env.get("ANDROID_VNDK_ROOT"),
                3, false));
        }
        
        return findings;
    }

    /**
     * Check for file system artifacts indicating AVD.
     */
    private static List<Finding> checkFileArtifacts() {
        List<Finding> findings = new ArrayList<>();
        
        // Common emulator artifact paths (relative to root)
        Set<String> artifactPaths = Set.of(
            "data/local/tmp/AndroidSdcard",
            "data/data/com.android.vndk.current.service/",
            "system/etc/vintf/manifest.xml"
        );
        
        for (String path : artifactPaths) {
            Path fullPath;
            
            // Try multiple base paths
            String[] bases = {"/", System.getProperty("user.dir")};
            for (String base : bases) {
                fullPath = new File(base + path).getAbsoluteFile();
                if (fullPath.exists()) {
                    findings.add(new Finding(
                        "FILE_ARTIFACT_FOUND",
                        "Emulator artifact found: " + fullPath.getAbsolutePath(),
                        3, false));
                }
            }
        }
        
        // Check for AVD-specific files in common locations
        if (System.getProperty("os.name").toLowerCase().contains("linux")) {
            File qemuPidFile = new File("/proc/self/status");
            if (qemuPidFile.exists()) {
                try (BufferedReader reader = Files.newBufferedReader(qemuPidFile.toPath())) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        if (line.startsWith("Name:") && 
                            (line.contains("qemu") || line.contains("emulator"))) {
                            findings.add(new Finding(
                                "QEMU_PROCESS_NAME",
                                "Process appears to be QEMU/Emulator: " + line,
                                4, false));
                        }
                    }
                } catch (IOException e) {
                    // Ignore read errors
                }
            }
        }
        
        return findings;
    }

    /**
     * Check running processes for emulator indicators.
     */
    private static List<Finding> checkProcesses() {
        List<Finding> findings = new ArrayList<>();
        
        if (System.getProperty("os.name").toLowerCase().contains("linux")) {
            try {
                // Read /proc/self/status to check process name
                File selfStatus = new File("/proc/self/status");
                if (selfStatus.exists()) {
                    try (BufferedReader reader = Files.newBufferedReader(
                            selfStatus.toPath())) {
                        String line;
                        while ((line = reader.readLine()) != null) {
                            if (line.startsWith("Name:") && 
                                (line.contains("qemu") || line.contains("emulator"))) {
                                findings.add(new Finding(
                                    "PROCESS_NAME_SIGNATURE",
                                    "Process name indicates emulator: " + line,
                                    4, false));
                            } else if (line.startsWith("VmRSS:") && 
                                       Long.parseLong(line.split(":")[1]) > 500 * 1024) {
                                // Large memory footprint - common in emulators
                                findings.add(new Finding(
                                    "HIGH_MEMORY_FOOTPRINT",
                                    "Process has large memory: " + line,
                                    2, true));
                            }
                        }
                    }
                }
            } catch (IOException e) {
                // Ignore read errors
            }
        }
        
        return findings;
    }

    /**
     * Check network interfaces for emulator patterns.
     */
    private static List<Finding> checkNetworkInterfaces() {
        List<Finding> findings = new ArrayList<>();
        
        if (System.getProperty("os.name").toLowerCase().contains("linux")) {
            try {
                // Read /proc/net/dev to check for virtual interfaces
                File netDevFile = new File("/proc/net/dev");
                if (netDevFile.exists()) {
                    try (BufferedReader reader = Files.newBufferedReader(
                            netDevFile.toPath())) {
                        String line;
                        boolean foundHeader = false;
                        
                        while ((line = reader.readLine()) != null) {
                            // Skip header lines
                            if (!foundHeader && !line.isEmpty() && 
                                !line.contains("Inter") && !line.contains("lo")) {
                                foundHeader = true;
                            }
                            
                            if (foundHeader && line.contains(":")) {
                                String[] parts = line.split(":");
                                String ifaceName = parts[0].trim();
                                
                                // Check for virtual interface patterns
                                if (ifaceName.toLowerCase().contains("veth") ||
                                    ifaceName.toLowerCase().contains("virbr") ||
                                    ifaceName.toLowerCase().contains("docker")) {
                                    
                                    findings.add(new Finding(
                                        "VIRTUAL_NETWORK_INTERFACE",
                                        "Virtual network interface detected: " + ifaceName,
                                        2, true));
                                }
                                
                                // Check for emulator-specific patterns
                                if (ifaceName.contains("emulator") || 
                                    ifaceName.contains("avd")) {
                                    findings.add(new Finding(
                                        "EMULATOR_NETWORK_INTERFACE",
                                        "Emulator network interface: " + ifaceName,
                                        3, false));
                                }
                            }
                        }
                    }
                }
            } catch (IOException e) {
                // Ignore read errors
            }
        }
        
        return findings;
    }

    /**
     * Check for sensor/data patterns that differ between real devices and emulators.
     */
    private static List<Finding> checkSensorPatterns(Map<String, String> env) {
        List<Finding> findings = new ArrayList<>();
        
        // Emulator sensors often have specific characteristics
        if (env.containsKey("ro.hardware.sensors")) {
            String sensorHardware = env.get("ro.hardware.sensors");
            
            // Common emulator sensor hardware strings
            Set<String> emulatorSensors = Set.of(
                "emulator",
                "sdk_phone",
                "generic"
            );
            
            if (sensorHardware.toLowerCase().contains("emulator") ||
                sensorHardware.toLowerCase().contains("sdk")) {
                findings.add(new Finding(
                    "EMULATOR_SENSOR_HARDWARE",
                    "Sensor hardware indicates emulator: " + sensorHardware,
                    3, false));
            }
        }
        
        // Check for simulated gyroscope/accelerometer patterns
        if (env.containsKey("ro.hardware.gyro")) {
            String gyro = env.get("ro.hardware.gyro");
            
            // Emulator often has "simulated" or similar markers
            if (gyro.toLowerCase().contains("simulated") ||
                gyro.toLowerCase().contains("emulated")) {
                findings.add(new Finding(
                    "SIMULATED_GYROSCOPE",
                    "Simulated gyroscope detected: " + gyro,
                    3, false));
            }
        }
        
        return findings;
    }

    /**
     * Result of detection.
     */
    public static class DetectionResult {
        private final int score;
        private final List<Finding> findings;
        private String verdict;
        
        public DetectionResult(int score, List<Finding> findings) {
            this.score = score;
            this.findings = findings;
            this.verdict = getVerdict(score);
        }

        public int getScore() {
            return score;
        }

        public String getVerdict() {
            return verdict;
        }

        public List<Finding> getFindings() {
            return Collections.unmodifiableList(findings);
        }

        @Override
        public String toString() {
            StringBuilder sb = new StringBuilder();
            sb.append("DetectionResult{score=").append(score)
              .append(", verdict='").append(verdict).append('\'')
              .append(", findingsCount=").append(findings.size()).append('}');
            
            if (!findings.isEmpty()) {
                sb.append("\n  Details:");
                for (Finding f : findings) {
                    sb.append("\n    - ").append(f);
                }
            }
            
            return sb.toString();
        }
    }

    /**
     * Individual finding with metadata.
     */
    public static class Finding {
        private final String type;
        private final String description;
        private final int weight;
        private final boolean isSoft; // Soft findings are less definitive
        
        public Finding(String type, String description, int weight, boolean isSoft) {
            this.type = type;
            this.description = description;
            this.weight = weight;
            this.isSoft = isSoft;
        }

        public String getType() {
            return type;
        }

        public String getDescription() {
            return description;
        }

        public int getWeight() {
            return weight;
        }

        public boolean isSoft() {
            return isSoft;
        }

        @Override
        public String toString() {
            return "  [" + (isSoft ? "[soft]" : "") + 
                   "] type=" + type + ", desc='" + description + "'";
        }
    }

    // ============================================================================
    // Runnable Demo / Main Entry Point
    // ============================================================================
    
    public static void main(String[] args) {
        System.out.println("=== AVD Spoof Detector ===\n");
        
        DetectionResult result = detect();
        
        System.out.println("Score: " + result.getScore() + "/10");
        System.out.println("Verdict: " + result.getVerdict());
        System.out.println("\nFindings (" + result.getFindings().size() + "):");
        
        if (result.getFindings().isEmpty()) {
            System.out.println("  [none]");
        } else {
            for (Finding f : result.getFindings()) {
                System.out.println(f);
            }
        }
        
        // Quick summary recommendation
        System.out.println("\n=== Summary ===");
        switch (result.getVerdict()) {
            case "LIKELY_EMULATOR":
                System.out.println("Device is likely running in an Android emulator.");
                break;
            case "POSSIBLE_EMULATOR":
                System.out.println("Some indicators suggest emulation, but confidence is moderate.");
                break;
            default:
                System.out.println("No strong evidence of emulation detected.");
                break;
        }

File: polyglot/go/emulator_avd_spoof_detector.go
package emulator_avd_spoof_detector

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

// EmulatorModel represents known AVD model signatures
type EmulatorModel struct {
	Name       string
	Pattern    *regexp.Regexp
	Score      int // Base score contribution (0-30)
	Description string
}

var emulatorModels = []EmulatorModel{
	{"Android SDK built for x86", regexp.MustCompile(`.*x86.*`), 25, "Classic AVD x86"},
	{"Generic", regexp.MustCompile(`.*Generic.*`), 20, "Generic AVD template"},
	{"Pixel Emulator", regexp.MustCompile(`.*Pixel.*Emulator.*`), 24, "Google Pixel emulator"},
	{"Nexus Emulator", regexp.MustCompile(`.*Nexus.*Emulator.*`), 23, "Nexus emulator variant"},
	{"Shamu", regexp.MustCompile(`.*shamu.*`), 18, "Pixel XL emulator codename"},
	{"Rex", regexp.MustCompile(`.*rex.*`), 17, "Pixel emulator codename"},
	{"Tuna", regexp.MustCompile(`.*tuna.*`), 16, "Pixel 2 emulator codename"},
	{"Blue", regexp.MustCompile(`.*blue.*`), 15, "Pixel 3 emulator codename"},
	{"Bluejay", regexp.MustCompile(`.*bluejay.*`), 14, "Pixel 3 XL emulator codename"},
}

// HardwareProfile represents expected hardware for real devices vs emulators
type HardwareProfile struct {
	ModelName     string
	Cores         int
	MemoryMB      int64
	IsEmulator    bool
	Score         int
	Description   string
}

var knownEmulatorHardware = []HardwareProfile{
	{"Android SDK built for x86", 4, 2048, true, 25, "Classic x86 AVD"},
	{"Generic", 4, 1024, true, 20, "Generic AVD profile"},
	{"Pixel Emulator", 8, 4096, true, 24, "Google Pixel emulator"},
}

// Detector holds state for a single detection run
type Detector struct {
	modelName     string
	buildFingerprints []string
	hardware       HardwareProfile
	score         int
	verdict       Verdict
	reasons       []string
}

// Verdict represents the final determination
type Verdict struct {
	Score        int    // 0-100, higher = more likely emulator
	Level        string // "CLEAN", "SUSPICIOUS", "LIKELY", "CONFIRMED"
	Confidence   float64 // 0.0-1.0
}

// Level constants for readability
const (
	VerdictClean    VerdictLevel = "CLEAN"
	VerdictSuspicious VerdictLevel = "SUSPICIOUS"
	VerdictLikely   VerdictLevel = "LIKELY"
	VerdictConfirmed VerdictLevel = "CONFIRMED"
)

// Level thresholds
var levelThresholds = []struct {
	Score    int
	Level    VerdictLevel
	Confidence float64
}{
	{0, VerdictClean, 1.0},
	{25, VerdictSuspicious, 0.75},
	{50, VerdictLikely, 0.50},
	{75, VerdictConfirmed, 0.25},
}

// Detect runs the full analysis and returns a Verdict
func (d *Detector) Detect() Verdict {
	d.score = 0
	d.reasons = make([]string, 0)

	// Check model name against known emulator patterns
	d.checkModelName()

	// Check hardware profile
	d.checkHardwareProfile()

	// Check build fingerprints
	d.checkBuildFingerprints()

	// Calculate confidence and determine level
	d.calculateConfidence()

	return d.verdict
}

func (d *Detector) checkModelName() {
	for _, model := range emulatorModels {
		if model.Pattern.MatchString(d.modelName) {
			d.score += model.Score
			d.reasons = append(d.reasons, fmt.Sprintf("Model name matches '%s' pattern", model.Name))
		}
	}

	if d.score == 0 && !strings.Contains(strings.ToLower(d.modelName), "emulator") {
		d.reasons = append(d.reasons, "Model name does not match known emulator patterns")
	}
}

func (d *Detector) checkHardwareProfile() {
	for _, profile := range knownEmulatorHardware {
		if d.hardware.ModelName == profile.ModelName && d.hardware.IsEmulator {
			d.score += profile.Score
			d.reasons = append(d.reasons, fmt.Sprintf("Hardware matches '%s' emulator profile", profile.Description))
		}
	}

	// Check for suspiciously clean hardware (too many cores for typical phone)
	if d.hardware.Cores > 8 && !strings.Contains(strings.ToLower(d.modelName), "x86") {
		d.score += 10
		d.reasons = append(d.reasons, "High core count suggests x86 emulator or high-end device")
	}

	// Check for suspiciously low memory (common in emulators)
	if d.hardware.MemoryMB < 512 && !strings.Contains(strings.ToLower(d.modelName), "x86") {
		d.score += 8
		d.reasons = append(d.reasons, "Low memory profile common in AVDs")
	}
}

func (d *Detector) checkBuildFingerprints() {
	emulatorProps := map[string]int{
		"ro.product.model":           15,
		"ro.build.fingerprint":       12,
		"ro.bootloader.id":           10,
		"ro.hardware":                8,
	}

	for _, fp := range d.buildFingerprints {
		for prop, score := range emulatorProps {
			if strings.Contains(fp, prop) {
				d.score += score
				d.reasons = append(d.reasons, fmt.Sprintf("Build fingerprint contains '%s' indicator", prop))
			}
		}

		// Check for specific AVD fingerprint patterns
		avdPatterns := []string{
			"Android SDK built for x86",
			"com.android.emulator",
			"androidsdk/",
			"emulator-",
		}

		for _, pattern := range avdPatterns {
			if strings.Contains(fp, pattern) {
				d.score += 10
				d.reasons = append(d.reasons, fmt.Sprintf("Build fingerprint contains AVD indicator: '%s'", pattern))
			}
		}
	}

	if len(d.buildFingerprints) == 0 {
		d.reasons = append(d.reasons, "No build fingerprints available for analysis")
	}
}

func (d *Detector) calculateConfidence() {
	maxScore := 100
	if d.score > maxScore {
		d.score = maxScore
	}

	// Map score to level and confidence
	for _, threshold := range levelThresholds {
		if d.score >= threshold.Score {
			d.level = threshold.Level
			d.confidence = threshold.Confidence
			break
		}
	}

	// Adjust confidence based on number of reasons found
	reasonCount := len(d.reasons)
	if reasonCount > 3 {
		d.confidence += 0.15
	} else if reasonCount == 0 {
		d.confidence -= 0.20
	}

	if d.confidence > 1.0 {
		d.confidence = 1.0
	}
}

// NewDetector creates a new detector with the given hardware profile
func NewDetector(modelName string, cores int, memoryMB int64) *Detector {
	return &Detector{
		modelName: modelName,
		hardware: HardwareProfile{
			ModelName:  modelName,
			Cores:      cores,
			MemoryMB:   memoryMB,
			IsEmulator: false, // Will be set by checkHardwareProfile
		},
	}
}

// DetectFromProps analyzes build properties from a map
func (d *Detector) DetectFromProps(props map[string]string) Verdict {
	d.buildFingerprints = make([]string, 0)

	for key, value := range props {
		if !strings.Contains(key, "ro.") {
			continue // Skip non-RO properties
		}

		// Include relevant RO properties
		relevantProps := []string{
			"product.model",
			"build.fingerprint",
			"bootloader.id",
			"hardware",
			"device.product.name",
			"ro.build.version.release",
		}

		for _, rp := range relevantProps {
			if strings.Contains(key, rp) {
				d.buildFingerprints = append(d.buildFingerprints, fmt.Sprintf("%s=%s", key, value))
			}
		}

		// Also include the full property string for pattern matching
		d.buildFingerprints = append(d.buildFingerprints, value)
	}

	return d.Detect()
}

// DetectFromEnv analyzes environment variables (useful when running in containerized environments)
func (d *Detector) DetectFromEnv(env map[string]string) Verdict {
	d.buildFingerprints = make([]string, 0)

	for key, value := range env {
		if strings.Contains(strings.ToLower(key), "android") ||
			strings.Contains(strings.ToLower(key), "sdk") ||
			strings.Contains(strings.ToLower(key), "emulator") {
			d.buildFingerprints = append(d.buildFingerprints, fmt.Sprintf("%s=%s", key, value))
		}

		if strings.Contains(value, "android") ||
			strings.Contains(value, "sdk") ||
			strings.Contains(value, "emulator") {
			d.buildFingerprints = append(d.buildFingerprints, value)
		}
	}

	return d.Detect()
}

// DetectFromPath analyzes a file path (useful for /proc/self/cmdline or similar)
func (d *Detector) DetectFromPath(path string) Verdict {
	d.buildFingerprints = make([]string, 0)

	if path == "" {
		return d.Detect()
	}

	content, err := os.ReadFile(path)
	if err != nil {
		d.reasons = append(d.reasons, fmt.Sprintf("Failed to read file: %v", err))
		return d.Detect()
	}

	lines := strings.Split(string(content), "\n")
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if len(line) == 0 || !strings.Contains(strings.ToLower(line), "android") &&
			!strings.Contains(strings.ToLower(line), "emulator") {
			continue
		}

		d.buildFingerprints = append(d.buildFingerprints, line)
	}

	return d.Detect()
}

// DetectFromStream analyzes a reader (useful for /proc/self/cmdline or similar streams)
func (d *Detector) DetectFromStream(r *bufio.Reader) Verdict {
	d.buildFingerprints = make([]string, 0)

	buf := make([]byte, 4096)
	for {
		n, err := r.Read(buf)
		if n > 0 {
			lines := strings.Split(string(buf[:n]), "\n")
			for _, line := range lines {
				line = strings.TrimSpace(line)
				if len(line) == 0 || !strings.Contains(strings.ToLower(line), "android") &&
					!strings.Contains(strings.ToLower(line), "emulator") {
					continue
				}

				d.buildFingerprints = append(d.buildFingerprints, line)
			}
		}
		if err != nil || n == 0 {
			break
		}
	}

	return d.Detect()
}

// DetectFromReader wraps a reader for convenience
func (d *Detector) DetectFromReader(r io.Reader) Verdict {
	reader := bufio.NewReader(r)
	return d.DetectFromStream(reader)
}

// GetReport returns a human-readable report of the detection
func (d *Detector) GetReport() string {
	var sb strings.Builder

	sb.WriteString(fmt.Sprintf("=== AVD SPOOF DETECTION REPORT ===\n\n"))
	sb.WriteString(fmt.Sprintf("Model Name: %s\n", d.modelName))
	sb.WriteString(fmt.Sprintf("Hardware Profile:\n  Cores: %d\n  Memory: %d MB\n\n", d.hardware.Cores, d.hardware.MemoryMB))

	if len(d.buildFingerprints) > 0 {
		sb.WriteString(fmt.Sprintf("Build Fingerprints Analyzed (%d):\n", len(d.buildFingerprints)))
		for i, fp := range d.buildFingerprints[:min(10, len(d.buildFingerprints))] {
			sb.WriteString(fmt.Sprintf("  %d. %s\n", i+1, fp))
		}
		if len(d.buildFingerprints) > 10 {
			sb.WriteString(fmt.Sprintf("  ... and %d more\n", len(d.buildFingerprints)-10))
		}
		sb.WriteString("\n")
	}

	sb.WriteString(fmt.Sprintf("Score: %d/100\n", d.score))
	sb.WriteString(fmt.Sprintf("Level: %s\n", d.level))
	sb.WriteString(fmt.Sprintf("Confidence: %.2f\n\n", d.confidence))

	if len(d.reasons) > 0 {
		sb.WriteString("Reasons:\n")
		for i, reason := range d.reasons[:min(10, len(d.reasons))] {
			sb.WriteString(fmt.Sprintf("  %d. %s\n", i+1, reason))
		}
		if len(d.reasons) > 10 {
			sb.WriteString(fmt.Sprintf("  ... and %d more\n", len(d.reasons)-10))
		}
	}

	return sb.String()
}

// DetectFromReader is a convenience function that wraps the detector lifecycle
func DetectFromReader(r io.Reader) Verdict {
	d := NewDetector("", 4, 2048) // Default values
	return d.DetectFromStream(bufio.NewReader(r))
}

// DetectFromProps is a convenience function for build properties
func DetectFromProps(props map[string]string) Verdict {
	d := NewDetector("", 4, 2048)
	return d.DetectFromProps(props)
}

// DetectFromEnv is a convenience function for environment variables
func DetectFromEnv(env map[string]string) Verdict {
	d := NewDetector("", 4, 2048)
	return d.DetectFromEnv(env)
}

// DetectFromFile is a convenience function for file paths
func DetectFromFile(path string) (Verdict, error) {
	if path == "" {
		path = "/proc/self/cmdline"
	}

	d := NewDetector("", 4, 2048)
	reader, err := os.Open(path)
	if err != nil {
		return d.Detect(), fmt.Errorf("failed to open file: %w", err)
	}
	defer reader.Close()

	return d.DetectFromReader(reader), nil
}

// DetectFromStdin is a convenience function for stdin input
func DetectFromStdin() Verdict {
	d := NewDetector("", 4, 2048)
	return d.DetectFromStream(bufio.NewReader(os.Stdin))
}

// RunDemo runs a self-test with various inputs and prints results
func RunDemo() {
	fmt.Println("=== ROOTSENTRY: AVD SPOOF DETECTOR DEMO ===\n")

	// Test 1: Real device-like profile
	fmt.Println("--- Test 1: Real Device Profile ---")
	d1 := NewDetector("Pixel 6 Pro", 8, 4096)
	d1.buildFingerprints = []string{
		"ro.product.model=Pixel 6 Pro",
		"ro.build.fingerprint=google/pixel6/pro:12/SP1A.210812.023/X51S:user/release-keys",
	}
	v1 := d1.Detect()
	fmt.Printf("Score: %d, Level: %s, Confidence: %.2f\n\n", v1.Score, v1.Level, v1.Confidence)

	// Test 2: Classic AVD x86
	fmt.Println("--- Test 2: Classic AVD x86 ---")
	d2 := NewDetector("Android SDK built for x86", 4, 2048)
	d2.buildFingerprints = []string{
		"ro.product.model=Android SDK built for x86",
		"ro.build.fingerprint=com.android.emulator/Android SDK built for x86/12.0.0:user/release-keys",
	}
	v2 := d2.Detect()
	fmt.Printf("Score: %d, Level: %s, Confidence: %.2f\n\n", v2.Score, v2.Level, v2.Confidence)

	// Test 3: Generic AVD
	fmt.Println("---